File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [-1] * len(numbers)
    reNumbers = numbers[::-1]
    right = []
    idx = len(numbers) - 1
    max = 0
    
    while idx > -1:
        target = numbers.pop()
        reNumbers.pop()
        
        if len(numbers) > 1 and reNumbers == numbers:
            break;
            
        if max < target:
            max = target
        else:
            for i in range(len(right) - 1, -1, -1):
                if right[i] > target:
                    answer[idx] = right[i]
                    break
                
                right.pop()
                    
                

        right.append(target)
        idx -= 1
    return answer

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        Deque<Integer> set = new ArrayDeque<>();
        int[] answer = new int[numbers.length];
        Arrays.fill(answer, -1);
        for (int i = numbers.length -1 ; i > -1 ; i--) {
            while (!set.isEmpty() && numbers[i] >= set.peekLast() ){
                set.pollLast();
            }
            if (!set.isEmpty()){
                answer[i] = set.peekLast();
            }

            set.addLast(numbers[i]);
        }
        
        return answer;
    }
}